### PR TITLE
Prepare and show gRPC usage in CLI

### PR DESCRIPTION
**Pull request by kradalby**
This PR prepares and demonstrates the upcoming work to get the CLI communicating with the server via gRPC instead of modifying the database.

It sets up:
- Authentication (currently deny everything but localhost)
- Prepare for API keys to be used as initial authentication for gRPC/HTTP api for clients
- Lint and restructure proto files after `buf`'s suggestions
- Add a demonstration of create, delete and list namespaces with gRPC
  - Those were the easiest commands that would also use the integration tests.
 

I need help testing this in the wild, please help me set this up on servers and actually test that it behaves like you would expect. 

- Use the CLI to create namespaces with a running server
- test with and without Certs (lets encrypt and others)
- anything I could have forgotten.

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -208,6 +221,102 @@ func (h *Headscale) watchForKVUpdatesWorker() {
...
+	handler grpc.UnaryHandler) (interface{}, error) {
+
+	// Check if the request is coming from the on-server client.
+	// This is not secure, but it is to maintain maintainability
+	// with the "legacy" database-based client
+	// It is also neede for grpc-gateway to be able to connect to
+	// the server
+	p, _ := peer.FromContext(ctx)
+
+	if IsLocalhost(p.Addr.String()) {
```
Comment (juanfont) [inline app.go:236]: Would this be a problem with a reverse proxy? It could be shown as coming from localhost...

Comment (kradalby) [reply app.go:236]: Good point, I will make a config option to turn that off. 

That would be a problem while the api key part isn't ready tho...

Comment (ties) [reply app.go:236]: This feels like a use case for unix sockets and then optionally tcp (with some bearer token or mutual tls)

Comment (kradalby) [reply app.go]: its done

**Review by cure — COMMENTED** [review]
I've tried running this version and experimented with the namespaces cli commands. Both client and server are running this branch.

namespaces list => works
namespaces create => timeout:

<pre>
2021-10-31T14:37:01Z DBG HEADSCALE_ADDRESS environment is not set, connecting to unix socket. socket=/var/run/headscale.sock
2021-10-31T14:37:01Z TRC .../headscale/cmd/headscale/cli/utils.go:371 > Connecting via gRPC address=/var/run/headscale.sock
2021-10-31T14:37:06Z FTL Could not connect: context deadline exceeded error="context deadline exceeded"
</pre>

The unix socket file exists.

namespaces rename => works
namespaces destroy => not tested yet

File: cmd/headscale/cli/utils.go [hunk 2]
```diff
@@ -304,6 +316,69 @@ func getHeadscaleApp() (*headscale.Headscale, error) {
 	return h, nil
 }
 
+func getHeadscaleGRPCClient() (apiV1.HeadscaleServiceClient, *grpc.ClientConn) {
+	// TODO(kradalby): Make configurable
```
[the remark this answers is not in the document]
Comment (kradalby) [reply cmd/headscale/cli/utils.go:320]: Yes, that makes sense, will do.

**Comment by kradalby** [comment]
> I've tried running this version and experimented with the namespaces cli commands. Both client and server are running this branch.
> 
> namespaces list => works 
> namespaces rename => works 
> namespaces destroy => not tested yet
These should work as they will be moved in an upcoming PR

> namespaces create => timeout:
> 
> 2021-10-31T14:37:01Z DBG HEADSCALE_ADDRESS environment is not set, connecting to unix socket. socket=/var/run/headscale.sock
> 2021-10-31T14:37:01Z TRC .../headscale/cmd/headscale/cli/utils.go:371 > Connecting via gRPC address=/var/run/headscale.sock
> 2021-10-31T14:37:06Z FTL Could not connect: context deadline exceeded error="context deadline exceeded"
> The unix socket file exists.

hmm that is interesting, I will try to replicate locally, I got it working in docker and a simple "unrestricted" machine.

Could you check if it could be a permission issue with the socket? Either the socket itself, or if it has some systemd (or similar) limitations?

> Related: after I switched to this branch as the server, my tailscale clients are now all listed as offline in `headscale nodes list`, and the tailscale logs of the clients say
> 
> Oct 31 10:32:28 clientname tailscaled[3491991]: Received error: PollNetMap: Post "https://x.x.x.x/machine/REDACTED/map": http: server gave HTTP response to HTTPS client

Will investigate this as well

File: cmd/headscale/cli/utils.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kradalby) [reply cmd/headscale/cli/utils.go]: Fixed, that makes a lot more sense.

**Comment by kradalby** [comment]
> Related: after I switched to this branch as the server, my tailscale clients are now all listed as offline in `headscale nodes list`, and the tailscale logs of the clients say
> 
> Oct 31 10:32:28 clientname tailscaled[3491991]: Received error: PollNetMap: Post "https://x.x.x.x/machine/REDACTED/map": http: server gave HTTP response to HTTPS client

I think this should be sorted.

> 2021-10-31T14:37:01Z DBG HEADSCALE_ADDRESS environment is not set, connecting to unix socket. socket=/var/run/headscale.sock
> 2021-10-31T14:37:01Z TRC .../headscale/cmd/headscale/cli/utils.go:371 > Connecting via gRPC address=/var/run/headscale.sock
> 2021-10-31T14:37:06Z FTL Could not connect: context deadline exceeded error="context deadline exceeded"
The unix socket file exists.

I managed to replicate this, now I just need to figure out why.

**Comment by kradalby** [comment]
> 2021-10-31T14:37:01Z DBG HEADSCALE_ADDRESS environment is not set, connecting to unix socket. socket=/var/run/headscale.sock
> 2021-10-31T14:37:01Z TRC .../headscale/cmd/headscale/cli/utils.go:371 > Connecting via gRPC address=/var/run/headscale.sock
> 2021-10-31T14:37:06Z FTL Could not connect: context deadline exceeded error="context deadline exceeded"
> The unix socket file exists.

This should be addressed now. Please have another try @cure.

It was caused by the socket using the same TLS settings as the network socket.

**Comment by cure** [comment]
> > 2021-10-31T14:37:01Z DBG HEADSCALE_ADDRESS environment is not set, connecting to unix socket. socket=/var/run/headscale.sock
> > 2021-10-31T14:37:01Z TRC .../headscale/cmd/headscale/cli/utils.go:371 > Connecting via gRPC address=/var/run/headscale.sock
> > 2021-10-31T14:37:06Z FTL Could not connect: context deadline exceeded error="context deadline exceeded"
> > The unix socket file exists.
> 
> This should be addressed now. Please have another try @cure.
> 
> It was caused by the socket using the same TLS settings as the network socket.

Hmm, now the headscale service won't start anymore for me:

<pre>
systemd[1]: Started Headscale.
headscale[1185995]: panic: listen unix /var/run/headscale.sock: bind: address already in use
headscale[1185995]: goroutine 1 [running]:
headscale[1185995]: github.com/juanfont/headscale.(*Headscale).Serve(0xc0003b6580, 0x0, 0x0)
headscale[1185995]:         /personal/projects/headscale-krdalby/headscale/app.go:352 +0x1f66
headscale[1185995]: github.com/juanfont/headscale/cmd/headscale/cli.glob..func22(0x199f740, 0x1a636a8, 0x0, 0x0)
headscale[1185995]:         /personal/projects/headscale-krdalby/headscale/cmd/headscale/cli/server.go:25 +0xa5
headscale[1185995]: github.com/spf13/cobra.(*Command).execute(0x199f740, 0x1a636a8, 0x0, 0x0, 0x199f740, 0x1a636a8)
headscale[1185995]:         /personal/projects/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:860 +0x2c2
headscale[1185995]: github.com/spf13/cobra.(*Command).ExecuteC(0x199ed40, 0x0, 0x1a329c0, 0x0)
headscale[1185995]:         /personal/projects/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
headscale[1185995]: github.com/spf13/cobra.(*Command).Execute(...)
headscale[1185995]:         /personal/projects/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:902
headscale[1185995]: github.com/juanfont/headscale/cmd/headscale/cli.Execute()
headscale[1185995]:         /personal/projects/headscale-krdalby/headscale/cmd/headscale/cli/root.go:25 +0x31
headscale[1185995]: main.main()
headscale[1185995]:         /personal/projects/headscale-krdalby/headscale/cmd/headscale/headscale.go:80 +0x2cb
</pre>

The `/var/run/headscale.sock` file exists and there is no other headscale process running.

**Comment by kradalby** [comment]
> The `/var/run/headscale.sock` file exists and there is no other headscale process running.

Does it start if you clear out / delete the sock file before you start it?

I cant seem to get that behaviour when running with systemd, but I get it in macOS when running for dev purposes. 

I am not sure what is go should clear it up vs things like systemd.

**Comment by kradalby** [comment]
Added handling for the socket, please try again, but you might have to remove the socket if it exist already.

**Comment by cure** [comment]
> > The `/var/run/headscale.sock` file exists and there is no other headscale process running.
> 
> Does it start if you clear out / delete the sock file before you start it?
> 
> I cant seem to get that behaviour when running with systemd, but I get it in macOS when running for dev purposes.
> 
> I am not sure what is go should clear it up vs things like systemd.

You called it! Manually removing the socket file before starting headscale resolved the problem.

I can also confirm that the issue with `headscale namespaces create` is fixed. `headscale namespaces destroy` also works. And the issue with nodes listing offline is also gone.

**Comment by kradalby** [comment]
Ok, after looking around the internet, go does not seem to have any automatic handling, so I have added a routine waiting for different system calls to clear up the socket before we close the process. If we have other errors, we might have to clean up manually.

This should bring it into ready shape :)

**Comment by kradalby** [comment]
@cure Do you feel happy about this? Would you be up for approving? :)

**Review by cure — APPROVED** [review]
Yeah, I think it's good to merge, thanks!

**Pull request merged**